File: dart/lib/src/parser/parser.dart
import 'clause.dart';
import 'combinators.dart';
import 'match_result.dart';
import 'memo_entry.dart';

/// The squirrel parser: a memoizing recursive descent (packrat) parser that
/// directly supports left recursive PEG grammars.
class Parser {
  final Map<String, Clause> rules;
  final Set<String> transparentRules;
  final String topRuleName;
  final String input;
  final Map<Clause, Map<int, MemoEntry>> _memoTable = {};

  /// Records how many times a left recursive cycle has been expanded at each
  /// input position ("cycleDepthForPos" in the paper). A match attempt may be
  /// made past the end of the input, hence size input.length + 1.
  final List<int> memoVersion;

  Parser({required Map<String, Clause> rules, required this.topRuleName, required this.input})
      : rules = {},
        transparentRules = {},
        memoVersion = List.filled(input.length + 1, 0) {
    // Process rules: strip '~' prefix indicating a transparent rule
    for (final entry in rules.entries) {
      if (entry.key.startsWith('~')) {
        final ruleName = entry.key.substring(1);
        this.rules[ruleName] = entry.value;
        transparentRules.add(ruleName);
      } else {
        this.rules[entry.key] = entry.value;
      }
    }
  }

  /// Match a rule's top clause at a position, using memoization.
  ///
  /// Memoization is applied only at the granularity of rules (this method is
  /// only reached via [Ref.match] or the top-level [parse] call); clauses
  /// within a rule's clause tree recurse directly without memoization.
  MatchResult match(Clause clause, int pos) {
    if (pos > input.length) return mismatch;

    var memoEntry = _memoTable.putIfAbsent(clause, () => {}).putIfAbsent(pos, MemoEntry.new);
    return memoEntry.match(this, clause, pos);
  }

  /// Match a named rule at a position.
  MatchResult matchRule(String ruleName, int pos) {
    final clause = rules[ruleName];
    if (clause == null) {
      throw ArgumentError('Rule "$ruleName" not found');
    }
    return match(clause, pos);
  }

  /// Get the MemoEntry for a clause at a position (if it exists).
  MemoEntry? getMemoEntry(Clause clause, int pos) {
    return _memoTable[clause]?[pos];
  }

  /// The approximate position of the first syntax error: the largest input
  /// position of any mismatch recorded in the memo table, or -1 if there is
  /// none. (See the paper: the location of the first syntax error can
  /// generally be identified by searching the memo table for the largest
  /// position of any mismatch.)
  int syntaxErrorPosition() {
    var maxPos = -1;
    for (final entriesForClause in _memoTable.values) {
      for (final entry in entriesForClause.entries) {
        if (entry.value.result?.isMismatch == true && entry.key > maxPos) {
          maxPos = entry.key;
        }
      }
    }
    return maxPos;
  }

  /// Parse the input, starting by matching the top rule at position 0.
  ///
  /// The core algorithm performs no error recovery. If the top rule does not
  /// match the whole input, [ParseResult.hasSyntaxErrors] is set, and any
  /// unmatched trailing input is wrapped in a [SyntaxError] node.
  ParseResult parse() {
    final result = matchRule(topRuleName, 0);
    final hasSyntaxErrors = result.isMismatch || result.len != input.length;
    return ParseResult(
      input: input,
      // If the top rule didn't match at all, return a SyntaxError spanning the entire input
      root: !result.isMismatch ? result : SyntaxError(pos: 0, len: input.length),
      // Save the name of the top rule
      topRuleName: topRuleName,
      // Record which rules are transparent for AST construction
      transparentRules: transparentRules,
      hasSyntaxErrors: hasSyntaxErrors,
      // If matched only part of the input, create an additional SyntaxError for the unmatched input
      unmatchedInput: !result.isMismatch && result.len < input.length
          ? SyntaxError(pos: result.len, len: input.length - result.len)
          : null,
    );
  }
}

// -----------------------------------------------------------------------------------------------------------------

/// The result of parsing the input.
class ParseResult {
  /// The input string that was parsed.
  final String input;

  /// The top-level MatchResult obtained from parsing the input.
  final MatchResult root;

  /// The rule name of the toplevel match
  final String topRuleName;

  /// The set of transparent rules in the grammar (rules that do not generate AST or CST nodes).
  final Set<String> transparentRules;

  /// True if the top rule failed to match the entire input.
  final bool hasSyntaxErrors;

  /// Contains a SyntaxError for any unmatched input if the whole input was not matched.
  final SyntaxError? unmatchedInput;

  ParseResult(
      {required this.input,
      required this.root,
      required this.topRuleName,
      required this.transparentRules,
      required this.hasSyntaxErrors,
      this.unmatchedInput});

  /// Get the syntax errors from the parse.
  List<SyntaxError> getSyntaxErrors() {
    if (!hasSyntaxErrors) {
      return [];
    }
    var errors = <SyntaxError>[];
    if (root is SyntaxError) {
      errors.add(root as SyntaxError);
    }
    if (unmatchedInput != null) {
      errors.add(unmatchedInput!);
    }
    return errors;
  }
}

File: java/src/main/java/com/squirrelparser/parser/Parser.java
package com.squirrelparser.parser;

import static com.squirrelparser.parser.MatchResult.mismatch;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

import com.squirrelparser.clause.Clause;
import com.squirrelparser.clause.nonterminal.Ref;

/**
 * The squirrel parser with bounded error recovery.
 */
public final class Parser {
    private final Map<String, Clause> rules;
    private final Set<String> transparentRules;
    private final String topRuleName;
    private final String input;
    private final Map<Clause, Map<Integer, MemoEntry>> memoTable;
    private final int[] memoVersion;
    private boolean inRecoveryPhase = false;

    public Parser(Map<String, Clause> rules, String topRuleName, String input) {
        this.rules = new HashMap<>();
        this.transparentRules = new HashSet<>();
        this.topRuleName = topRuleName;
        this.input = input;
        this.memoTable = new HashMap<>();
        this.memoVersion = new int[input.length() + 1];

        // Process rules: strip '~' prefix indicating a transparent rule
        for (var entry : rules.entrySet()) {
            if (entry.getKey().startsWith("~")) {
                String ruleName = entry.getKey().substring(1);
                this.rules.put(ruleName, entry.getValue());
                transparentRules.add(ruleName);
            } else {
                this.rules.put(entry.getKey(), entry.getValue());
            }
        }
    }

    public Map<String, Clause> rules() { return rules; }
    public Set<String> transparentRules() { return transparentRules; }
    public String topRuleName() { return topRuleName; }
    public String input() { return input; }
    public int[] memoVersion() { return memoVersion; }
    public boolean inRecoveryPhase() { return inRecoveryPhase; }

    /**
     * Match a clause at a position, using memoization.
     */
    public MatchResult match(Clause clause, int pos, Clause bound) {
        if (pos > input.length()) {
            return mismatch();
        }

        // C5 (Ref Transparency): Don't memoize Ref independently
        if (clause instanceof Ref) {
            return clause.match(this, pos, bound);
        }

        @SuppressWarnings("unused")
        MemoEntry memoEntry = memoTable
            .computeIfAbsent(clause, k -> new HashMap<>())
            .computeIfAbsent(pos, k -> new MemoEntry());
        return memoEntry.match(this, clause, pos, bound);
    }

    public MatchResult match(Clause clause, int pos) {
        return match(clause, pos, null);
    }

    /**
     * Match a named rule at a position.
     */
    public MatchResult matchRule(String ruleName, int pos) {
        Clause clause = rules.get(ruleName);
        if (clause == null) {
            throw new IllegalArgumentException("Rule \"" + ruleName + "\" not found");
        }
        return match(clause, pos);
    }

    /**
     * Get the MemoEntry for a clause at a position (if it exists).
     */
    public MemoEntry getMemoEntry(Clause clause, int pos) {
        var clauseMap = memoTable.get(clause);
        return clauseMap != null ? clauseMap.get(pos) : null;
    }

    /**
     * Probe: Temporarily switch out of recovery mode to check if clause can match.
     */
    public MatchResult probe(Clause clause, int pos) {
        boolean savedPhase = inRecoveryPhase;
        inRecoveryPhase = false;
        MatchResult result = match(clause, pos);
        inRecoveryPhase = savedPhase;
        return result;
    }

    /**
     * Enable recovery mode (Phase 2).
     */
    public void enableRecovery() {
        inRecoveryPhase = true;
    }

    /**
     * Check if clause can match non-zero characters at position.
     */
    public boolean canMatchNonzeroAt(Clause clause, int pos) {
        MatchResult result = probe(clause, pos);
        return !result.isMismatch() && result.len() > 0;
    }

    /**
     * Parse input with two-phase error recovery.
     */
    public ParseResult parse() {
        // Phase 1: Discovery (try to parse without recovery from syntax errors)
        MatchResult result = matchRule(topRuleName, 0);
        boolean hasSyntaxErrors = result.isMismatch() || result.pos() != 0 || result.len() != input.length();
        if (hasSyntaxErrors) {
            // Phase 2: Attempt to recover from syntax errors
            enableRecovery();
            result = matchRule(topRuleName, 0);
        }

        return new ParseResult(
            input,
            !result.isMismatch() ? result : new SyntaxError(0, input.length()),
            topRuleName,
            transparentRules,
            hasSyntaxErrors,
            hasSyntaxErrors && result.len() < input.length()
                ? new SyntaxError(result.len(), input.length() - result.len())
                : null
        );
    }
}
